src/mpi-base.jl

```diff
@@ -525,6 +525,62 @@ function Reduce{T}(object::T, op::Op, root::Integer, comm::Comm)
     isroot ? recvbuf[1] : nothing
 end
 
+# T and T2 are different for MPI_IN_PLACE as the sendbuf
+function Allreduce{T}(sendbuf::MPIBuffertype{T}, recvbuf::MPIBuffertype{T},
+                      count::Integer, op::Op, comm::Comm)
+    flag = Ref{Cint}()
+    ccall(MPI_ALLREDUCE, Void, (Ptr{T}, Ptr{T}, Ptr{Cint}, Ptr{Cint}, Ptr{Cint},
+          Ptr{Cint}, Ptr{Cint}), sendbuf, recvbuf, &Int32(count), &mpitype(T), 
+          &op.val, &comm.val, flag)
+
+    if flag[] != 0
+      throw(ErrorException("Allreduce returned non-zero exit status"))
+    end
+
+    recvbuf
+end
+                  
+function Allreduce{T}(sendbuf::MPIBuffertype{T}, recvbuf::MPIBuffertype{T},
+                      op::Op, comm::Comm)
+    Allreduce(sendbuf, recvbuf, length(recvbuf), op, comm)
+end
+
+function Allreduce{T}(obj::T, op::Op, comm::Comm)
+    objref = Ref(obj)
+    outref = Ref{T}()
+    Allreduce(objref, outref, 1, op, comm)
+
+    outref[]
+end
+
+#=
+# in-place version
+function Allreduce!{T}(recvbuf::MPIBuffertype{T}, op::Op, comm::Comm)
+
+#  Allreduce(sendbuf, recvbuf, length(recvbuf), op, comm)
+    flag = zero(Cint)
+    ccall(MPI_ALLREDUCE, Void, (Ptr{Cint}, Ptr{Void}, Ptr{Cint}, Ptr{Cint}, 
+          Ptr{Cint}, Ptr{Cint}, Ptr{Cint}), &MPI_IN_PLACE, recvbuf, 
```

Comment:
`MPI_IN_PLACE` should be passed in stead of a pointer, not as content of the send buffer. In other words, `MPI_IN_PLACE` is a pointer, not an integer, in the sense that its address is important, not its value. Thus the way we capture its value is wrong; we need to capture its value instead. Let me add a comment there as well.